### PR TITLE
ungoogled-chromium: init at 81.0.4044.92-2

### PR DESCRIPTION
**Pull request by squalus**
###### Motivation for this change
Adds support for the [ungoogled-chromium](https://github.com/Eloston/ungoogled-chromium) patch set to the nixpkgs chromium derivation. Requested in https://github.com/NixOS/nixpkgs/issues/53579.

> ungoogled-chromium is Google Chromium, sans dependency on Google web services. It also features some tweaks to enhance privacy, control, and transparency (almost all of which require manual activation or enabling). ungoogled-chromium retains the default Chromium experience as closely as possible.

- Applies the full patch set (except for Widevine removal, which is covered by an existing patch)
- Runs all ungoogled-chromium patching utilities, including binary pruning and domain substitution.
- Applies the GN flags from ungoogled-chromium. Unforunately they are duplicated in the chromium nix expression. An alternative would be to parse them out of the flags.gn file in the ungoogled-chromium repo. That would still involve some tweaking and deduping though. I'm interested in feedback as to whether it's worth doing this.
- Supports multiple versions of chromium with the `ungoogled-src.nix` file, which maps a chromium version to a ungoogled-chromium revision (although it's not able to be auto-updated like `upstream-info.nix`).
- Tested 79.0.3945.79 on NixOS 19.09. Everything works ok for me, including WebGL acceleration.

The patch set can be enabled with the following nixpkgs config:

```
chromium = {
     ungoogled = true;
};
```

I'm very interested in feedback!

Prior work: https://github.com/NixOS/nixpkgs/pull/51195.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions (Gentoo)
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @bendlas @ivan @thefloweringash

### CODE REVIEW

File: pkgs/applications/networking/browsers/chromium/ungoogled.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{ stdenv
+, fetchFromGitHub
+, python3Packages
+, makeWrapper
+, patch
+}:
+{ rev, sha256 }:
+stdenv.mkDerivation {
```
Comment (worldofpeace) [inline pkgs/applications/networking/browsers/chromium/ungoogled.nix]: This expression lacks good whitespace like other expressions in nixpkgs.

See like https://github.com/NixOS/nixpkgs/blob/master/pkgs/desktops/pantheon/apps/elementary-camera/default.nix

Comment (squalus) [reply pkgs/applications/networking/browsers/chromium/ungoogled.nix]: Updated whitespace.

**Review by ivan — CHANGES_REQUESTED** [review]
If you would like to package ungoogled-chromium, I think it should be its own package, with the upstream versions decoupled from chromium's versions. Chromium is already frequently difficult and time-consuming (as in, whole day down the drain) to update. I cannot hold back a new release with 50 CVE fixes just because the ungoogled-chromium patchset has not been updated yet. This would also burden the chromium maintainer with another 3+ hour build and smoke test.

ungoogled-chromium is basically a separate browser product, even if it is distributed as a patchset, so I recommend copying nixpkgs chromium/ into a new directory and replacing the maintainers :-)

**Comment by squalus** [comment]
I'd be willing to copy/paste into a new package. I agree that the chromium release should not be held back due to a delay in the patchset release.

**Comment by squalus** [comment]
I updated the PR to duplicate the chromium expressions.

The maintenance workflow will be to start with copy/pasted chromium expressions, then reapply the expression diff on top of that, then update the ungoogled-chromium sources as needed.

Ideally I wouldn't need to modify the copy/pasted chromium expressions at all, instead just naively duplicate them and keep the patchset-related stuff in separate files. However, I couldn't figure out any way to do that with the current chromium expressions. What I would need is the ability to externally (through an override maybe) provide extra gnFlags and add extra postPatch steps.

**Comment by LouisDK1** [comment]
@squalus Any update on your work on a separate package? I'd love to have Ungoogled chromium available for NixOS.

**Comment by squalus** [comment]
@LouisDK1 Yeah, in the latest update to this PR there's a separate `chromiumUngoogled` package available. I've been running it for about a week.

**Comment by LouisDK1** [comment]
> @LouisDK1 Yeah, in the latest update to this PR there's a separate `chromiumUngoogled` package available. I've been running it for about a week.

Awesome. I differently think this should get into the main repo.
Wouldn't it be a good idea to  add "enableParallelBuilding = true;" somewhere in common.nix to minimize build time?

**Comment by LouisDK1** [comment]
@squalus - Please update package to: 79.0.3945.130-2 and if possible change the headline to match the fact that this is a separate package.

**Comment by squalus** [comment]
@LouisDK1 Updated the title, and left the version as-is. For now I'm aiming for the smallest possible diff from the "upstream" chromium nix expressions. I'll leave the testing and stabilization of new versions up to them.

**Comment by LouisDK1** [comment]
@ivan He have done it. Could you review and maybe accept the package? :)

**Review by worldofpeace — CHANGES_REQUESTED** [review]
Can we not call it `chromiumUngoogled` it's really backwards?
`ungoogled-chromium` makes more sense https://repology.org/project/ungoogled-chromium/versions.
(and please rename the directories)

**Comment by LouisDK1** [comment]
@squalus ping

**Comment by squalus** [comment]
- changed package name and directories to `ungoogled-chromium`
- updated version to 79.0.3945.117-1

**Review by worldofpeace — COMMENTED** [review]
I've built ungoogled-chromium and it works :sparkles:

**Comment by rycee** [comment]
@worldofpeace Do you feel that the changes you requested were sorted? Your review is still marked as requesting changes 🙂

**Comment by wizeman** [comment]
Works for me :+1: Thanks!

**Review by rycee — APPROVED** [review]
LGTM

**Comment by worldofpeace** [comment]
> @worldofpeace Do you feel that the changes you requested were sorted? Your review is still marked as requesting changes slightly_smiling_face

I don't feel I could properly commentate on adding another chromium flavor in nixpkgs other than it works. I don't completely approve of the maintainability of completely duplicating the chromium expressions. But I would be happy to use this.

**Comment by rycee** [comment]
@worldofpeace Yeah, agreed. It would be great if some more of the common parts could be abstracted out of these packages. I'm inclined to merge as-is, though, since I suppose that would be a job too large for this PR.

**Comment by worldofpeace** [comment]
> @worldofpeace Yeah, agreed. It would be great if some more of the common parts could be abstracted out of these packages. I'm inclined to merge as-is, though, since I suppose that would be a job too large for this PR.

It was actually suggested https://github.com/NixOS/nixpkgs/pull/76082#pullrequestreview-335436810.
I have no disagreements if you merge, I just hope to not see the expressions decay.

File: pkgs/applications/networking/browsers/ungoogled-chromium/update.nix [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+      SHA1 hash.
+
+      Using this method, we can distinguish whether the URL is available or
+      whether it's not based on the actual content.
+
+      So let's use tryEval as soon as it's working with fetchurl in Nix.
+    */
+    tryFetch = url: let
+      # SHA1 hash collisions from https://shattered.io/static/shattered.pdf:
+      collisions = runCommand "sha1-collisions" {
```
Comment (volth) [inline pkgs/applications/networking/browsers/ungoogled-chromium/update.nix]: That `April 1`'s collision-based update has been removed from `nixpkgs`, hasn't it?

Comment (squalus) [reply pkgs/applications/networking/browsers/ungoogled-chromium/update.nix]: It's still here: https://github.com/NixOS/nixpkgs/blob/08f68313a47a2093dc0f408a706b2c125bc59c95/pkgs/applications/networking/browsers/chromium/update.nix#L113

(this PR is just a copy of the chromium directory with whatever small changes are required to apply the ungoogled-chromium patchset)

**Comment by LouisDK1** [comment]
@worldofpeace  - can you "accept" your request as it has been fulfilled?
Anything else holding back this package from being merged?

**Comment by travankor** [comment]
:shipit:

**Comment by symphorien** [comment]
I compiled this PR successfully.

**Comment by shirona** [comment]
@worldofpeace the changes you requested were fixed. Could we get the patch merged, pretty please? :(

**Comment by primeos** [comment]
<details><summary>My delayed opinion (disclaimer: probably not really constructive, feel free to skip)</summary><p>

---

tbh I'm not sure if https://github.com/NixOS/nixpkgs/pull/76082#pullrequestreview-335436810 was a good idea (cop-pasting the Chromium expression instead of overriding it) because:
- I consider it a bad practice in general
- It might be more difficult to track the differences
  - But maybe not since we have good `diff` tooling :)
- Both expressions are likely to diverge over time if we aren't careful, resulting in unnecessary/unintended functional differences
- If `ungoogled-chromium` isn't actively maintained, I'll quickly fall behind and get insecure
  - E.g. currently it's already one patch release behind

But at least it doesn't make maintaining `chromium` more difficult than it already is :)

But I'm too late with this comment (sorry about this delayed note) and who knows, it might be better the way it is anyway.

---

</p>
</details>

@squalus I'd suggest the following to maintain this (I'm the current `chromium` maintainer):
- Regularly check for differences, e.g. with:
  `git diff --no-index pkgs/applications/networking/browsers/{,ungoogled-}chromium`
  - E.g. currently there are VA-API, `flashplayer-ppapi`, and version differences
  - Re-using changes from `chromium` will likely help you avoid build failures (I generally try to fix them in advance via testing the `beta` and `dev` channels as well)
- I can ping/notify you on `chromium` PRs if you want
- If possible ping/notify me about changes that make sense for `chromium` as well so I can re-use them
  - I.e. we'll try to "cherry-pick" changes in both directions
- Maybe find a second maintainer, e.g. for NixOS stable (backports) and as fallback

Also: Thanks for adding this package and good luck! I'm looking forward to a successful cooperation :)

**Comment by squalus** [comment]
Thanks for merging!

@primeos thanks for the comments. I have some maintenance scripts that will do this process semi-automatically. I'll commit them if I can make them look sensible. My plan for maintenance is just to periodically take the entire chromium directory, blindly copy it over, and apply the small set of ungoogled diffs. I would also like to work on the chromium expressions to make this process easier. :+1: 

No need to notify me over changes, I'll be periodically monitoring them (but feel free). I'm also looking forward to working with you!

**Comment by danielfullmer** [comment]
@squalus Also, I've been testing the chromium PRs by @primeos in the past. If interested, you can CC me on future update PRs of ungoogled-chromium and I can do similar basic building / testing. I have a beefy 32-core threadripper which makes this relatively quick for me.

**Comment by squalus** [comment]
@danielfullmer Will do. Thank you!

**Pull request merged**